commit 2c894cb3adb8f1c1f2653af05879a3e96d0add21
Author: Rafael Sales <rafaelcds@gmail.com>
Date:   Mon Oct 12 23:18:51 2015 -0300

    Backport #21950 :
    
    Fix generated projection fields in group by query
    
    Closes #21922
    
    Let `Book(id, author_id)`, `Photo(id, book_id, author_id)` and `Author(id)`
    
    Running `Book.group(:author_id).joins(:photos).count` will produce:
    
    * Rails 4.2 - conflicts `author_id` in both projection and group by:
    ```sql
    SELECT COUNT(*) AS count_all, author_id AS author_id
      FROM "books" INNER JOIN "photos" ON "photos"."book_id" = "books"."id"
     GROUP BY author_id
    ```
    
    * Master (9d02a25) - conflicts `author_id` only in projection:
    ```sql
    SELECT COUNT(*) AS count_all, author_id AS author_id
      FROM "books" INNER JOIN "photos" ON "photos"."book_id" = "books"."id"
     GROUP BY "books"."author_id"
    ```
    
    * With this fix:
    ```sql
    SELECT COUNT(*) AS count_all, "books"."author_id" AS books_author_id
      FROM "books" INNER JOIN "photos" ON "photos"."book_id" = "books"."id"
     GROUP BY "books"."author_id"
    ```

diff --git a/activerecord/lib/active_record/relation/calculations.rb b/activerecord/lib/active_record/relation/calculations.rb
index 3fe5bfd..6f466ab 100644
--- a/activerecord/lib/active_record/relation/calculations.rb
+++ b/activerecord/lib/active_record/relation/calculations.rb
@@ -287,6 +287,7 @@ module ActiveRecord
       else
         group_fields = group_attrs
       end
+      group_fields = arel_columns(group_fields)
 
       group_aliases = group_fields.map { |field|
         column_alias_for(field)
